File: Python/mergesort.py
import time

# Merge function
def merge(left, right):
    merged = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1

    # Append remaining elements
    merged.extend(left[i:])
    merged.extend(right[j:])
    return merged

# Merge Sort function
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)


# ---------------------------- MAIN PROGRAM ----------------------------
if __name__ == "__main__":
    n = int(input("Enter number of elements: "))
    print("Enter elements separated by space:")
    arr = list(map(int, input().split()))

    start_time = time.perf_counter()
    sorted_arr = merge_sort(arr)
    end_time = time.perf_counter()

    print("\nSorted array:")
    print(sorted_arr)

    print("\nExecution time:", end_time - start_time, "seconds")

File: Rust/MergeSort.rs
use std::io::{self, Read};
use std::time::Instant;

// Merge two sorted slices
fn merge(left: &[i32], right: &[i32]) -> Vec<i32> {
    let mut result = Vec::with_capacity(left.len() + right.len());
    let mut i = 0;
    let mut j = 0;

    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            result.push(left[i]);
            i += 1;
        } else {
            result.push(right[j]);
            j += 1;
        }
    }

    // Append remaining elements
    result.extend_from_slice(&left[i..]);
    result.extend_from_slice(&right[j..]);

    result
}

// Merge sort function
fn merge_sort(arr: &[i32]) -> Vec<i32> {
    if arr.len() <= 1 {
        return arr.to_vec();
    }

    let mid = arr.len() / 2;
    let left = merge_sort(&arr[..mid]);
    let right = merge_sort(&arr[mid..]);

    merge(&left, &right)
}

fn main() {
    // Read all input at once (fast and simple)
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut iter = input.split_whitespace();

    println!("Enter number of elements:");
    let n: usize = match iter.next() {
        Some(v) => v.parse().unwrap(),
        None => return,
    };

    println!("Enter elements separated by space:");
    let mut arr = Vec::with_capacity(n);
    for _ in 0..n {
        if let Some(v) = iter.next() {
            arr.push(v.parse::<i32>().unwrap());
        }
    }

    let start = Instant::now();
    let sorted = merge_sort(&arr);
    let duration = start.elapsed();

    println!("\nSorted array:");
    for x in &sorted {
        print!("{} ", x);
    }
    println!();

    println!(
        "\nExecution time: {:.9} seconds",
        duration.as_secs_f64()
    );
}
